akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala

```diff
@@ -0,0 +1,131 @@
+import akka.actor.testkit.typed.TestKitSettings
+import akka.actor.testkit.typed.scaladsl._
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.Behavior
+import akka.actor.typed.SupervisorStrategy
+import akka.actor.typed.Terminated
+import akka.actor.typed.scaladsl.ActorContext
+import akka.persistence.journal.inmem.InmemJournal
+import akka.persistence.query.journal.leveldb.scaladsl.LeveldbReadJournal
+import akka.persistence.query.EventEnvelope
+import akka.persistence.query.PersistenceQuery
+import akka.persistence.query.Sequence
+import akka.persistence.typed.ExpectingReply
+import akka.persistence.typed.PersistenceId
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl.Sink
+import com.typesafe.config.Config
+import com.typesafe.config.ConfigFactory
+import org.scalatest.WordSpecLike
+
+object PersistentBehaviorReplySpec {
+  def conf: Config = ConfigFactory.parseString(
+    s"""
+    akka.loglevel = INFO
+    # akka.persistence.typed.log-stashing = INFO
+    akka.persistence.journal.leveldb.dir = "target/typed-persistence-${UUID.randomUUID().toString}"
+    akka.persistence.journal.plugin = "akka.persistence.journal.leveldb"
+    akka.persistence.snapshot-store.plugin = "akka.persistence.snapshot-store.local"
+    akka.persistence.snapshot-store.local.dir = "target/typed-persistence-${UUID.randomUUID().toString}"
+    """)
+
+  sealed trait Command[ReplyMessage] extends ExpectingReply[ReplyMessage]
+  final case class IncrementWithConfirmation(override val replyTo: ActorRef[Done]) extends Command[Done]
+  final case class IncrementReplyLater(override val replyTo: ActorRef[Done]) extends Command[Done]
+  final case class ReplyNow(override val replyTo: ActorRef[Done]) extends Command[Done]
+  final case class GetValue(replyTo: ActorRef[State]) extends Command[State]
+
+  sealed trait Event
+  final case class Incremented(delta: Int) extends Event
+
+  final case class State(value: Int, history: Vector[Int])
+
+  def counter(persistenceId: PersistenceId)(implicit system: ActorSystem[_]): Behavior[Command[_]] =
+    Behaviors.setup(ctx ⇒ counter(ctx, persistenceId))
+
+  def counter(
+    ctx:           ActorContext[Command[_]],
+    persistenceId: PersistenceId): PersistentBehavior[Command[_], Event, State] = {
+    PersistentBehavior.withEnforcedReplies[Command[_], Event, State](
+      persistenceId,
+      emptyState = State(0, Vector.empty),
+      commandHandler = (state, cmd) ⇒ cmd match {
+
+        case cmd: IncrementWithConfirmation ⇒
+          Effect.persist(Incremented(1))
+            .thenReplyDone(cmd)
+
+        case cmd: IncrementReplyLater ⇒
+          Effect.persist(Incremented(1))
+            .thenRun((_: State) ⇒ ctx.self ! ReplyNow(cmd.replyTo))
+            .thenNoReply(cmd)
+
+        case cmd: ReplyNow ⇒
+          Effect.replyDone(cmd)
+
+        case query: GetValue ⇒
+          Effect.reply(query)(state)
+
```

Comment:
note that this only compiles if a `ReplyEffect` is returned, which can be created with `Effect.reply`, `Effect.noReply`, `Effect.replyDone`, `.thenReply`, `.thenNoReply`, `.thenReplyDone`.

Reply:
Looks like a message based implementation of phantom types .. nice!

Reply:
lol, I have always looked for a use case for those phantoms

Reply:
How will it work if I want the reply to come from another actor?

Also, `.noReply` and `.thenNoReply` seem like odd options given that we are trying to force a reply.

Reply:
Exactly, it's conscious opt-out decision. Reason could be things like:
* not all commands require a reply
* a reply is sent later in the process, e.g. after some interaction with another actor

Reply:
> How will it work if I want the reply to come from another actor?

In the end it's just about sending a message to the `replyTo: ActorRef` in the command, so you can carry on that to another actor or keep it (or the original command) around until it's time to reply.

That is somewhat illustrated in this example with the `IncrementReplyLater` that is sending a `ReplyNow` to self, which is then sending the reply.

Reply:
When I read this: `withEnforcedReplies` that seems to imply to me that we are enforcing that a reply happen. Looking at it from the outside I would be tempted to assume that if someone has implemented `withEnforcedReplies` then I can always expect that I will receive a reply.

That's obviously not true here. In fact, `withEnforcedReplies` actually doesn't give you any more guarantees about whether a reply is being returned than you had when it wasn't there. The only thing it is doing is making it slightly harder to write bad code. Although, if you are likely to forget to reply, then you are probably also likely to forget to include the `withEnforcedReplies` trait. There is maybe a small amount of additional security once it's there, but there is still a lot of room to mess things up, and the trait adds additional complexity (new ways to reply etc).

So I guess we have to ask what our goal is here. Is our goal to make it harder to introduce a bug in the code? If that's the case, then we see a small amount of success with this. If our goal is to provide a more intent revealing API, or one that is simpler and easier to use, then I think this fails to meet that criteria.

Reply:
The primary goal was to help was to detect that reply was forgotten, avoid mistakes.

I didn't put much thinking into the name `withEnforcedReplies`, so open for suggestions.

> small amount of additional security once it's there

In what way is it a small amount? It will not compile if you don't.

Reply:
It's small because there is still ways to mess it up. Assuming I remember to put the trait in, I could still use the wrong reply type (eg. replyDone instead of reply etc.). But the real reason I say it is small is because I am not sure that forgetting the reply is a huge problem. It's something that happens on occasion, but likely gets caught relatively quickly.

This is honestly not a problem I see a lot of people complaining about. I am not sure I have encountered it on an engagement. I have probably done it a few times in my own code, but then realized my mistake very quickly and fixed it.

I will say though that I don't see many support cases. So maybe it's a problem that comes up often on support cases.

Reply:
Ah yes, I see now that you can't in fact use the wrong reply type. Thanks.

If forgetting the reply is a huge problem in Lagom, it's not one that I have seen. So as I said, it must be something that comes up in support cases.